test/logic_values_test.dart

```diff
@@ -11,23 +11,794 @@
+        equals(LogicValues.fromString('1111'))
+      );
+      expect( // srl
+        LogicValues.fromString('1111') >>> 2,
+        equals(LogicValues.fromString('0011'))
+      );
+    });
+
+  });
+
+  group('comparison operations', () {
+
+    test('equality', () {
+      expect( // == equal
+        LogicValues.fromString('1111')  == LogicValues.fromString('1111'),
+        equals(true)
+      );
+      expect( // == not equal
+        LogicValues.fromString('1111')  == LogicValues.fromString('1110'),
+        equals(false)
+      );
+      expect( // eq equal
+        LogicValues.fromString('1111').eq(LogicValues.fromString('1111')),
+        equals(LogicValue.one)
+      );
+      expect( // eq not equal, valid
+        LogicValues.fromString('1111').eq(LogicValues.fromString('1110')),
+        equals(LogicValue.zero)
+      );
+      expect( // eq not equal, invalid
+        LogicValues.fromString('1111').eq(LogicValues.fromString('111x')),
+        equals(LogicValue.x)
+      );
+    });
+
+    test('greater', () {
+      expect( // >
+        LogicValues.fromString('0111') > LogicValues.fromString('0110'),
+        equals(LogicValue.one)
+      );
+      expect( // not >
+        LogicValues.fromString('0111') > LogicValues.fromString('0111'),
+        equals(LogicValue.zero)
+      );
+      expect( // >=
+        LogicValues.fromString('0111') >= LogicValues.fromString('0111'),
+        equals(LogicValue.one)
+      );
+      expect( // not >=
+        LogicValues.fromString('0110') >= LogicValues.fromString('0111'),
+        equals(LogicValue.zero)
+      );
+      expect( // x involved
+        LogicValues.fromString('0110') >= LogicValues.fromString('011x'),
+        equals(LogicValue.x)
+      );
+      /*
+      expect( // mismatched lengths
+        LogicValues.fromString('0110') >= LogicValues.fromString('011000'),
+        throwsA(isA<Exception>())
+      );
+      */
+    });
+
+    test('less', () {
+      expect( // <
+        LogicValues.fromString('0111') < 8,
+        equals(LogicValue.one)
+      );
+      expect( // not <
+        LogicValues.fromString('0111') < 7,
+        equals(LogicValue.zero)
+      );
+      expect( // <=
+        LogicValues.fromString('0111') <= 7,
+        equals(LogicValue.one)
+      );
+      expect( // not <=
+        LogicValues.fromString('0110') <= 5,
+        equals(LogicValue.zero)
+      );
+      expect( // x involved
+        LogicValues.fromString('011x') <= 10,
+        equals(LogicValue.x)
+      );
+    });
+
+  });
+
+  group('arithmetic operations', () {
+
+    test('addsub', () {
+      expect( // + normal
+        LogicValues.fromString('0001') + LogicValues.fromString('0011'),
+        equals(LogicValues.fromString('0100')) // 1 + 3 = 4
+      );
+      expect( // - normal
+        LogicValues.fromString('0001') - LogicValues.fromString('0001'),
+        equals(LogicValues.fromString('0000')) // 1 - 1 = 0
+      );
+      /*
+      expect( // + overflow
```

Comment:
these tests failed due to a bug in _SmallLogicValues, nice job!
they appear to be the same at first inspection since the toString only shows 4 bits, but the internal representation was, for example `10000`, and the equality did not mask.
same bug exists in _BigLogicValues, but these tests won't catch it.  add similar tests with >64 bits as well?

Reply:
filed #22